File: src/main/java/net/hellomouse/alexscavesenriched/utils/CodecUtils.java
package net.hellomouse.alexscavesenriched.utils;

import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.mojang.serialization.Codec;
import com.mojang.serialization.DataResult;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.util.ExtraCodecs;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.Ingredient;

import java.util.Optional;
import java.util.function.Function;

public class CodecUtils {
    // net.minecraft.item.ItemStack.CODEC is _weird_, count is required, oh wait, no, its `Count` not `count` ...
    public static final Codec<ItemStack> ITEMSTACK_CODEC = RecordCodecBuilder.create(
            instance -> instance.group(
                            BuiltInRegistries.ITEM.byNameCodec().fieldOf("item").forGetter(ItemStack::getItem),
                            Codec.INT.optionalFieldOf("count", 1).forGetter(ItemStack::getCount),
                            CompoundTag.CODEC.optionalFieldOf("tag").forGetter(arg -> Optional.ofNullable(arg.getTag()))
                    )
                    .apply(instance, (item, integer, compoundTag) -> new ItemStack(item, integer, compoundTag.orElse(null)))
    );
    public static Codec<Ingredient> INGRIDIENT_CODEC = ExtraCodecs.JSON.comapFlatMap(tryDataResult(Ingredient::fromJson), Ingredient::toJson);

    private static <T> Function<JsonElement, DataResult<T>> tryDataResult(Function<JsonElement, T> converter) {
        return (elm) -> {
            DataResult<T> result;
            try {
                result = DataResult.success(converter.apply(elm));
            } catch (JsonParseException e) {
                result = DataResult.error(e::getMessage);
            }
            return result;
        };
    }
}

File: src/main/kotlin/net/hellomouse/alexscavesenriched/utils/CodecUtils.kt
package net.hellomouse.alexscavesenriched.utils

import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonSyntaxException
import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.ExtraCodecs
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.Ingredient
import java.util.*
import kotlin.jvm.optionals.getOrNull

object CodecUtils {
    @JvmField
    val INGREDIENT_CODEC: Codec<Ingredient> =
        ExtraCodecs.JSON.comapFlatMap(
            tryDataResult(Ingredient::fromJson),
            Ingredient::toJson
        )

    // net.minecraft.item.ItemStack.CODEC is _weird_, count is required, oh wait, no, its `Count` not `count` ...
    @JvmField
    val ITEM_STACK_CODEC: MapCodec<ItemStack> = RecordCodecBuilder.mapCodec { instance ->
        instance.group(
            BuiltInRegistries.ITEM.byNameCodec().fieldOf("item")
                .forGetter(ItemStack::getItem),
            Codec.INT.optionalFieldOf("count", 1).forGetter(ItemStack::getCount),
            CompoundTag.CODEC.optionalFieldOf("tag")
                .forGetter { itemStack -> Optional.ofNullable(itemStack.tag) }
        ).apply(instance) { item, int, nbt -> ItemStack(item, int, nbt.getOrNull()) }
    }

    private fun <T> tryDataResult(converter: (JsonElement) -> T): (JsonElement) -> DataResult<T> {
        return { jsonElement ->
            try {
                DataResult.success(converter(jsonElement))
            } catch (e: JsonSyntaxException) {
                DataResult.error { e.message }
            } catch (e: JsonParseException) {
                DataResult.error { e.message }
            }
        }
    }
}